existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,292 @@
+If you want a theoretical proof of this statement, you can check this [tread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).
+
+### Usage
+
+In order to proceed further, we'll have to enable the `RankNTypes` extension. This lets us write higher-rank types.
+
+By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)
+
+For example, this function has a rank-1 type:
+
+```haskell
+f :: forall a b. a -> b -> a
+
+-- Or, to make the precedence explicit:
+f :: forall a b. (a -> b -> a)
+```
+
+With the `RankNTypes` extension, we can write rank-2 types - types where a `forall` appears to the left of one arrow:
+
+```haskell
+f :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+
+Because `forall a` appears to the left of one arrow, `a` is said to be existentially quantified.
+It is also true for every `forall` that is introduced to the left of an odd number of functional arrows.
+But what does this mean?
+
+In contrast to the universal quantifier, the existential one means that the function is itself responsible for instantiating what a type variable should mean. So the result type of existentially quantified variable will become known when we address it at the 'definition-site' (i.e., in the function body).
+This also means that no one can specify its type outside of function.
+
+Let's take a closer look on practical examples.
+
+We will start with functions:
+
+```haskell
+func :: forall a c. a -> (forall b. (Show b) => b -> IO ()) -> IO c
+```
+There are a few things to mention here.
+
+* In the type signature, `a` and `c` are universally quantified variables, and `b` is existentially quantified.
+* The same variable can't be quantified more than once.
+* You can see that constraints for `b` are declared _after_ `b` has been introduced with the `forall` quantifier
+* You can choose type for `b` only inside `func` definition, and you can't do it when you call `func` (we will discuss it further below).
+
+Here's a real-life example of using existential types in functions.
+Imagine a function that prints some logs while calculating the result.
+You may want to write logs into a console or into some file.
+In order to do it, you pass a logging function as an argument.
+
+```haskell
+func :: _ -> IO ()
+func log = do
+  username <- getUsername
+  log username
+
+  userCount <- getUserCount
+  log userCount
+
+getUsername :: IO String
+getUsername = undefined
+
+getUserCount :: IO Int
+getUserCount = undefined
+```
+
+What type should the log function have?
+Since it somehow logs both `String` and `Int`:
+
+```haskell
+log :: Show a => a -> IO ()
+```
+However, if we add this signature to `func` and try to call it with different arguments:
+
+```haskell
+func :: Show a => (a -> IO ()) -> IO ()
```

Comment:
In the previous sentance you say "if we add this signature to `func`". But if we do that, then the resulting function should be 
```
 func :: (Show a => (a -> IO ())) -> IO ()
```
since `Show a` constraint come from the `log` function.